jerry-core/include/jerryscript-core.h

```diff
@@ -195,15 +195,16 @@ typedef jerry_value_t (*jerry_vm_exec_stop_callback_t) (void *user_p);
 typedef bool (*jerry_object_property_foreach_t) (const jerry_value_t property_name,
                                                  const jerry_value_t property_value,
                                                  void *user_data_p);
-/**
- * Function type for user context allocation
- */
-typedef void *(*jerry_user_context_init_t) (void);
 
 /**
- * Function type for user context deallocation
+ * User context item manager
  */
-typedef void (*jerry_user_context_deinit_t) (void *user_context_p);
+typedef struct
+{
+  void (*init_cb) (void *); /**< callback responsible for initializing a context item */
```

Comment:
@zherczeg if you mean that it should be `void (*init_cb)(void *newly_allocated_p, void *hint_p)` then I don't think that's necessary, because if any user ever needs to use the context for two different reasons, they can just store a default reason in the context data during `init_cb ()`, and change the reason if it's wrong:
```c
/* this calls init_cb () */
my_data_t *my_data_p = jerry_get_context_data (&my_manager);
if (my_data_p->init_hint != MY_INIT_HINT)
{
  reinit_my_data(my_data_p, MY_INIT_HINT);
}
```

Or, they can have two different managers for different kinds of context data. That's what the [unit test](https://github.com/jerryscript-project/jerryscript/pull/1833/files#diff-a91107b501aafed86478d129d0a4d2e3) does.

I tend to think of context data as global static data that cannot be stored in a global static variable because it's different for each Jerry context, and the same compilation unit has to run in two different contexts. As such, I don't believe there should be any more degrees of freedom.

Now, given this analogy, global static variables can be initialized to non-trivial values without any hints, and in our case, the `init_cb` has full access to the JS world of its context, so it can definitely make decisions about how it should initialize itself.

Another implication of adding the hint is that it probably needs to be stored along with the `manager_p` in the header and then passed to `deinit_cb`.

At that point you're probably better off with having two different managers.